File: src/lib/annis/api/graphupdate.cpp
/*
   Copyright 2017 Thomas Krause <thomaskrause@posteo.de>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

#include "graphupdate.h"


using namespace annis::api;

GraphUpdate::GraphUpdate()
 : lastConsistentChangeID(0)
{
}

void GraphUpdate::addNode(std::string name, std::string type)
{
  std::shared_ptr<AddNodeEvent> evt = std::make_shared<AddNodeEvent>();
  evt->changeID = lastConsistentChangeID + diffs.size() + 1;
  evt->nodeName = name;
  evt->nodeType = type;

  diffs.push_back(evt);
}

void GraphUpdate::deleteNode(std::string name)
{
  std::shared_ptr<DeleteNodeEvent> evt = std::make_shared<DeleteNodeEvent>();
  evt->changeID = lastConsistentChangeID + diffs.size() + 1;
  evt->nodeName = name;

  diffs.push_back(evt);
}

void GraphUpdate::addNodeLabel(std::string nodeName, std::string ns, std::string name, std::string value)
{
  std::shared_ptr<AddNodeLabelEvent> evt = std::make_shared<AddNodeLabelEvent>();
  evt->changeID = lastConsistentChangeID + diffs.size() + 1;
  evt->nodeName = nodeName;
  evt->annoNs = ns;
  evt->annoName = name;
  evt->annoValue = value;

  diffs.push_back(evt);
}

void GraphUpdate::deleteNodeLabel(std::string nodeName, std::string ns, std::string name)
{
   std::shared_ptr<DeleteNodeLabelEvent> evt = std::make_shared<DeleteNodeLabelEvent>();
   evt->changeID = lastConsistentChangeID + diffs.size() + 1;
   evt->nodeName = nodeName;
   evt->annoNs = ns;
   evt->annoName = name;

   diffs.push_back(evt);
}

void GraphUpdate::addEdge(std::string sourceNode, std::string targetNode, std::string layer,
                          std::string componentType, std::string componentName)
{
   std::shared_ptr<AddEdgeEvent> evt = std::make_shared<AddEdgeEvent>();
   evt->changeID = lastConsistentChangeID + diffs.size() + 1;
   evt->sourceNode = sourceNode;
   evt->targetNode = targetNode;
   evt->layer = layer;
   evt->componentType = componentType;
   evt->componentName = componentName;

   diffs.push_back(evt);
}

void GraphUpdate::deleteEdge(std::string sourceNode, std::string targetNode, std::string layer,
                             std::string componentType, std::string componentName)
{
   std::shared_ptr<DeleteEdgeEvent> evt = std::make_shared<DeleteEdgeEvent>();
   evt->changeID = lastConsistentChangeID + diffs.size() + 1;
   evt->sourceNode = sourceNode;
   evt->targetNode = targetNode;
   evt->layer = layer;
   evt->componentType = componentType;
   evt->componentName = componentName;

   diffs.push_back(evt);
}

void GraphUpdate::addEdgeLabel(std::string sourceNode, std::string targetNode, std::string layer,
                               std::string componentType, std::string componentName,
                               std::string annoNs, std::string annoName, std::string annoValue)
{
    std::shared_ptr<AddEdgeLabelEvent> evt = std::make_shared<AddEdgeLabelEvent>();
    evt->changeID = lastConsistentChangeID + diffs.size() + 1;

    evt->sourceNode = sourceNode;
    evt->targetNode = targetNode;
    evt->layer = layer;
    evt->componentType = componentType;
    evt->componentName = componentName;

    evt->annoNs = annoNs;
    evt->annoName = annoName;
    evt->annoValue = annoValue;

    diffs.push_back(evt);

}

void GraphUpdate::deleteEdgeLabel(std::string sourceNode, std::string targetNode, std::string layer,
                                  std::string componentType, std::string componentName,
                                  std::string annoNs, std::string annoName)
{
   std::shared_ptr<DeleteEdgeLabelEvent> evt = std::make_shared<DeleteEdgeLabelEvent>();
   evt->changeID = lastConsistentChangeID + diffs.size() + 1;

   evt->sourceNode = sourceNode;
   evt->targetNode = targetNode;
   evt->layer = layer;
   evt->componentType = componentType;
   evt->componentName = componentName;

   evt->annoNs = annoNs;
   evt->annoName = annoName;

   diffs.push_back(evt);
}

void GraphUpdate::finish()
{
   if(!diffs.empty())
   {
      std::shared_ptr<UpdateEvent> lastEvent = *(diffs.rbegin());
      lastConsistentChangeID = lastEvent->changeID;
   }
}

bool GraphUpdate::isConsistent() const
{
   if(diffs.empty())
   {
      return true;
   }
   else
   {
      std::shared_ptr<UpdateEvent> lastEvent = *(diffs.rbegin());
      if(lastEvent)
      {
        return lastConsistentChangeID == lastEvent->changeID;
      }
      else
      {
        return false;
      }
   }
}

File: java-api/src/main/java/org/corpus_tools/graphannis/api/GraphUpdate.java
/*
 * Copyright 2018 Thomas Krause.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.corpus_tools.graphannis.api;

import org.corpus_tools.graphannis.capi.CAPI;

/**
 * An API for applying atomic updates to a graph DB.
 *    
 * @author Thomas Krause <thomaskrause@posteo.de>
 */
public class GraphUpdate {
    private final CAPI.AnnisGraphUpdate instance;

    public GraphUpdate() {
        this.instance = CAPI.annis_graphupdate_new();
    }

    public CAPI.AnnisGraphUpdate getInstance() {
        return this.instance;
    }

    public void addNode(String node_name, String node_type) {
        CAPI.annis_graphupdate_add_node(instance, node_name, node_type);
    }

    public void addNode(String node_name) {
        CAPI.annis_graphupdate_add_node(instance, node_name, "node");
    }

    public void deleteNode(String node_name) {
        CAPI.annis_graphupdate_delete_node(instance, node_name);
    }

    public void addNodeLabel(String node_name, String anno_ns, String anno_name, String anno_value) {
        CAPI.annis_graphupdate_add_node_label(instance, node_name, anno_ns, anno_name, anno_value);
    }

    public void deleteNodeLabel(String node_name, String anno_ns, String anno_name) {
        CAPI.annis_graphupdate_delete_node_label(instance, node_name, anno_ns, anno_name);
    }

    public void addEdge(String source_node, String target_node, String layer, String component_type,
            String component_name) {
        CAPI.annis_graphupdate_add_edge(instance, source_node, target_node, layer, component_type, component_name);
    }

    public void deleteEdge(String source_node, String target_node, String layer, String component_type,
            String component_name) {
        CAPI.annis_graphupdate_add_edge(instance, source_node, target_node, layer, component_type, component_name);
    }

    public void addEdgeLabel(String source_node, String target_node, String layer, String component_type,
            String component_name, String anno_ns, String anno_name, String anno_value) {
        CAPI.annis_graphupdate_add_edge_label(instance, source_node, target_node, layer, component_type, component_name,
                anno_ns, anno_name, anno_value);
    }

    public void deleteEdgeLabel(String source_node, String target_node, String layer, String component_type,
            String component_name, String anno_ns, String anno_name) {
        CAPI.annis_graphupdate_delete_edge_label(instance, source_node, target_node, layer, component_type,
                component_name, anno_ns, anno_name);
    }
}